### PR TITLE
fix changes overwriting of clients in the RBDClientMap

### PR DESCRIPTION
**Pull request by weirdwiz**
This commit restructures things to make the metric more reliable, it also keeps tracks of all the nodes.

### CODE REVIEW

**Comment by weirdwiz** [comment]
TODO: combine kube_state metrics and blocklist metric to create an alert

File: metrics/internal/cache/ceph-blocklist.go [hunk 1]
```diff
@@ -161,10 +159,15 @@ func runCephOSDBlocklist(config *cephMonitorConfig) ([]CephBlocklistLs, error) {
 
 	re := regexp.MustCompile(`\[[^\[\]]+\]`)
 	match := re.Find(cmd)
-	if len(match) == 0 && err != nil {
+	klog.Info(match)
```
Comment (umangachapagain) [inline metrics/internal/cache/ceph-blocklist.go:162]: We don't need to see it unless we're debugging. Can we remove it?

File: metrics/internal/cache/pv.go [hunk 2]
```diff
@@ -31,18 +31,20 @@ type PersistentVolumeStore struct {
 	Mutex sync.RWMutex
 	// Store is a map of PV UID to PersistentVolumeAttributes
 	Store             map[types.UID]PersistentVolumeAttributes
-	RBDClientMap      map[string]Clients
+	RBDClientMap      map[string][]string
```
Comment (umangachapagain) [inline metrics/internal/cache/pv.go:34]: Add a comment about what this map stores.

File: metrics/internal/collectors/ceph-blocklist.go [hunk 3]
```diff
@@ -6,28 +6,48 @@ import (
...
 type CephBlocklistCollector struct {
+	kubeClient            clientset.Interface
 	CephBlocklistStore    *internalcache.CephBlocklistStore
 	PersistentVolumeStore *internalcache.PersistentVolumeStore
 	// Metrics Descriptors
 	NodeRBDBlocklist *prometheus.Desc
 }
 
-func NewCephBlocklistCollector(blocklistStore *internalcache.CephBlocklistStore, pvStore *internalcache.PersistentVolumeStore) *CephBlocklistCollector {
+func NewCephBlocklistCollector(opts *options.Options, blocklistStore *internalcache.CephBlocklistStore, pvStore *internalcache.PersistentVolumeStore) *CephBlocklistCollector {
```
Comment (umangachapagain) [inline metrics/internal/collectors/ceph-blocklist.go:26]: `options.Options` should be the last parameter.

File: metrics/internal/collectors/ceph-blocklist.go [hunk 4]
```diff
@@ -6,28 +6,48 @@ import (
...
 		CephBlocklistStore:    blocklistStore,
 		PersistentVolumeStore: pvStore,
 		NodeRBDBlocklist: prometheus.NewDesc(
 			prometheus.BuildFQName(namespace, "", "rbd_client_blocklisted"),
-			"State of the rbd client on a node, 0 = Unblocked, 1 = Blocked", []string{"node_name"}, nil),
+			"State of the rbd client on a node, 0 = Unblocked, 1 = Blocked", []string{"node"}, nil),
 	}
 }
 
+func (c *CephBlocklistCollector) getNodes() (map[string]bool, error) {
```
Comment (umangachapagain) [inline metrics/internal/collectors/ceph-blocklist.go:37]: `getBlockedNodes` might be a better name for what this function does.

File: metrics/internal/collectors/ceph-blocklist.go [hunk 5]
```diff
@@ -45,21 +65,32 @@ func (c *CephBlocklistCollector) Collect(ch chan<- prometheus.Metric) {
...
 		}
 	}
+
+	for node, blocked := range blockedNodes {
+		if blocked {
+			ch <- prometheus.MustNewConstMetric(c.NodeRBDBlocklist, prometheus.GaugeValue, 1, node)
+		} else {
+			ch <- prometheus.MustNewConstMetric(c.NodeRBDBlocklist, prometheus.GaugeValue, 0, node)
+		}
+	}
```
Comment (umangachapagain) [inline metrics/internal/collectors/ceph-blocklist.go:93]: Do we need to run this loop? Would it be possible to create these metrics in the above loop when you're updating the `blockedNodes` map?

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: metrics/deploy/prometheus-ocs-rules.yaml [hunk 6]
```diff
@@ -194,12 +194,21 @@ spec:
     rules:
     - alert: ODFRBDClientBlocked
       annotations:
-        description: A kernel RBD client on node {{ $labels.node_name }} is blocklisted by Ceph for more than 10s. Please check Ceph logs.
-        message: A kernel RBD client on node {{ $labels.node_name }} is blocklisted by Ceph.
+        description: An RBD client might be blocklisted by Ceph on node {{ $labels.node_name }} for more than 10s. Please check pod description for more details.
```
Comment (umangachapagain) [inline metrics/deploy/prometheus-ocs-rules.yaml:197]: Description needs to be more clear that we're talking about pods with `CreateContainerError` error whose pod description has "read-only filesystem" error.

File: metrics/internal/cache/pv.go [hunk 7]
```diff
@@ -154,8 +154,6 @@ func getNodeNameForPV(pv *corev1.PersistentVolume, kubeClient clientset.Interfac
 		return "", fmt.Errorf("persistent volume %s is not bound to any claim", pv.Name)
 	}
 
-	klog.Info(pv.Name, pv.Spec.ClaimRef.Namespace)
-
```
Comment (umangachapagain) [inline metrics/internal/cache/pv.go:147]: shouldn't be in this commit.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: metrics/internal/collectors/ceph-blocklist.go (hunk 4 above)
Comment (weirdwiz) [reply metrics/internal/collectors/ceph-blocklist.go:37]: this only gets all the node on the system, and initializes all the values as false

File: metrics/deploy/prometheus-ocs-rules.yaml (hunk 6 above)
Comment (weirdwiz) [reply metrics/deploy/prometheus-ocs-rules.yaml:197]: ack, updating

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**